pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -287,111 +281,67 @@ public boolean isMutableSegment() {
+            case DOUBLE:
+              break;
+            default:
+              throw new UnsupportedOperationException(
+                  "Unsupported data type: " + dataType + " for MV no-dictionary column: " + column);
           }
         }
+      }
 
-        if (useNativeTextIndex) {
-          textIndex = new NativeMutableTextIndex(column);
-        } else {
+      // Null value vector
+      MutableNullValueVector nullValueVector = _nullHandlingEnabled ? new MutableNullValueVector() : null;
 
-          // TODO - this logic is in the wrong place and belongs in a Lucene-specific submodule,
-          //  it is beyond the scope of realtime index pluggability to do this refactoring, so realtime
-          //  text indexes remain statically defined. Revisit this after this refactoring has been done.
-          RealtimeLuceneTextIndex luceneTextIndex =
-              new RealtimeLuceneTextIndex(column, new File(config.getConsumerDir()), _segmentName, stopWordsInclude,
-                  stopWordsExclude);
-          if (_realtimeLuceneReaders == null) {
-            _realtimeLuceneReaders = new RealtimeLuceneIndexRefreshState.RealtimeLuceneReaders(_segmentName);
-          }
-          _realtimeLuceneReaders.addReader(luceneTextIndex);
-          textIndex = luceneTextIndex;
+      Map<IndexType, MutableIndex> mutableIndexes = new HashMap<>();
+      for (IndexType<?, ?, ?> indexType : IndexService.getInstance().getAllIndexes()) {
+        if (!specialIndexes.contains(indexType)) {
+          addMutableIndex(mutableIndexes, indexType, context, indexConfigs);
         }
-      } else {
-        textIndex = null;
       }
 
-      // Json index
-      JsonIndexConfig jsonIndexConfig = jsonIndexConfigs.get(column);
-      MutableJsonIndex jsonIndex =
-          jsonIndexConfig != null ? indexProvider.newJsonIndex(context.forJsonIndex(jsonIndexConfig)) : null;
-
-      // H3 index
-      // TODO consider making this overridable
-      MutableH3Index h3Index;
-      try {
-        H3IndexConfig h3IndexConfig = h3IndexConfigs.get(column);
-        h3Index = h3IndexConfig != null ? new MutableH3Index(h3IndexConfig.getResolution()) : null;
-      } catch (IOException e) {
-        throw new RuntimeException(String.format("Failed to initiate H3 index for column: %s", column), e);
+      // TODO - this logic is in the wrong place and belongs in a Lucene-specific submodule,
+      //  it is beyond the scope of realtime index pluggability to do this refactoring, so realtime
+      //  text indexes remain statically defined. Revisit this after this refactoring has been done.
+      MutableIndex textIndex = mutableIndexes.get(StandardIndexes.text());
+      if (textIndex instanceof RealtimeLuceneTextIndex) {
+        if (_realtimeLuceneReaders == null) {
+          _realtimeLuceneReaders = new RealtimeLuceneIndexRefreshState.RealtimeLuceneReaders(_segmentName);
+        }
+        _realtimeLuceneReaders.addReader((RealtimeLuceneTextIndex) textIndex);
```

Comment:
This is the poor man's way to add the index to lucene readers now. The TODO was originally created by Richard (although it has been moved and now git attributes o me) and still applies. This logic doesn't belong here, but this should be treated in a better way once we decide to invest resources in reducing this hack